localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -692,6 +711,67 @@ bool NDTScanMatcher::validate_converged_param(
   return is_ok_converged_param;
 }
 
+void NDTScanMatcher::estimate_covariance(
+  const pclomp::NdtResult & ndt_result, const Eigen::Matrix4f & initial_pose_matrix,
+  std::array<double, 36> & ndt_covariance, const rclcpp::Time & sensor_ros_time)
+{
+  // Laplace approximation
+  const Eigen::Matrix2d hessian_inv = -ndt_result.hessian.inverse().block(0, 0, 2, 2);
+  const Eigen::SelfAdjointEigenSolver<Eigen::Matrix2d> lap_eigensolver(hessian_inv);
+  Eigen::Matrix2d rot = Eigen::Matrix2d::Identity();
+  if (lap_eigensolver.info() == Eigen::Success) {
+    const Eigen::Vector2d lap_eigen_vec = lap_eigensolver.eigenvectors().col(1);
+    const double th = std::atan2(lap_eigen_vec.y(), lap_eigen_vec.x());
+    rot = Eigen::Rotation2Dd(th);
+  }
+
+  // first result is added to mean and covariance
+  const Eigen::Vector2d p2d(ndt_result.pose(0, 3), ndt_result.pose(1, 3));
+  Eigen::Vector2d tmp_centroid = p2d;
+  Eigen::Matrix2d tmp_cov = p2d * p2d.transpose();
+
+  geometry_msgs::msg::PoseArray multi_ndt_result_msg;
+  geometry_msgs::msg::PoseArray multi_initial_pose_msg;
+  multi_ndt_result_msg.header.stamp = sensor_ros_time;
+  multi_ndt_result_msg.header.frame_id = map_frame_;
+  multi_initial_pose_msg.header.stamp = sensor_ros_time;
+  multi_initial_pose_msg.header.frame_id = map_frame_;
+  multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(ndt_result.pose));
+  multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(initial_pose_matrix));
+
+  // multiple searches
+  for (const auto & offset_vec : offset_array_) {
+    const Eigen::Vector2d offset_2d = rot * offset_vec;
+
+    Eigen::Matrix4f sub_initial_pose_matrix(Eigen::Matrix4f::Identity());
+    sub_initial_pose_matrix = ndt_result.pose;
+    sub_initial_pose_matrix(0, 3) += static_cast<float>(offset_2d.x());
+    sub_initial_pose_matrix(1, 3) += static_cast<float>(offset_2d.y());
+
+    auto sub_output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
+    ndt_ptr_->align(*sub_output_cloud, sub_initial_pose_matrix);
+    const Eigen::Matrix4f sub_ndt_result = ndt_ptr_->getResult().pose;
+
+    const Eigen::Vector2d sub_p2d = sub_ndt_result.topRightCorner<2, 1>().cast<double>();
+    tmp_centroid += sub_p2d;
+    tmp_cov += sub_p2d * sub_p2d.transpose();
+
+    multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(sub_ndt_result));
+    multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
+  }
+  const Eigen::Vector2d centroid = tmp_centroid / (offset_array_.size() + 1);
+
+  const Eigen::Matrix2d pca_covariance =
+    (tmp_cov / (offset_array_.size() + 1) - (centroid * centroid.transpose()));
+  ndt_covariance[0] = output_pose_covariance_[0] + pca_covariance(0, 0);
+  ndt_covariance[1] = pca_covariance(1, 0);
+  ndt_covariance[6] = pca_covariance(0, 1);
+  ndt_covariance[7] = output_pose_covariance_[7] + pca_covariance(1, 1);
```

Comment:
This code assumes that the roll/pitch/yaw of `ndt_covariance` is initialized elsewhere. Since this is risking becoming a bug, how about the following?
```
ndt_covariance = output_pose_covariance_;
ndt_covariance[0 + 6 * 0] += pca_covariance(0, 0);
ndt_covariance[1 + 6 * 0] += pca_covariance(1, 0);
ndt_covariance[0 + 6 * 1] += pca_covariance(0, 1);
ndt_covariance[1 + 6 * 1] += pca_covariance(1, 1);
```